# Assurance of Learning (Number 3)

## Import Library

In [14]:
# Import Library

import math
import numpy as np
from scipy.optimize import newton, bisect

### Data

In [15]:
# Bag production data

data = [
    1863, 1614, 2570, 1685, 2101, 1811, 2457, 2171, 2134, 2502, 2358, 2399, 2048, 2523, 2086, 2391,
    2150, 2340, 3129, 2277, 2964, 2997, 2747, 2862, 3405, 2677, 2749, 2755, 2963, 3161, 3623, 2768,
    3141, 3439, 3601, 3531, 3477, 3376, 4027, 3175, 3274, 3334, 3964, 3649, 3502, 3688, 3657, 4422,
    4197, 4441, 4736, 4521, 4485, 4644, 5036, 4876, 4789, 4544, 4975, 5211, 4880, 4933, 5079, 5339,
    5232, 5520, 5714, 5260, 6110, 5334, 5988, 6235, 6365, 6266, 6345, 6118, 6497, 6278, 6638, 6590,
    6271, 7246, 6584, 6594, 7092, 7326, 7409, 7976, 7959, 8012, 8195, 8008, 8313, 7791, 8368, 8933,
    8756, 8613, 8705, 9098, 8769, 9544, 9050, 9186, 10012, 9685, 9966, 10048, 10244, 10740, 10318,
    10393, 10986, 10635, 10731, 11749, 11849, 12123, 12274, 11666, 11960, 12629, 12915, 13051,
    13387, 13309, 13732, 13162, 13644, 13808, 14101, 13992, 15191, 15018, 14917, 15046, 15556,
    15893, 16388, 16782, 16716, 17033, 16896, 17689
]

# From Month 1 to Month 144
months = np.arange(1, 145)

### Newton Raphson Function

In [16]:
# f --> f(x), g --> g(x)
# x0 --> Initial guess
# e -> error
# max_iter --> Maximum iteration

def netwon_raphson (f, g, x0, e = 0.01, max_iter = 15):
    for i in range (1, max_iter):
        x1 = x0 - (f(x0) / g (x0))
    
        if np.abs (f(x1) < e):
            return x1
        
        x0 = x1

    return None

### Bisection Method

In [17]:
def bisection (f, a, b, e=0.0001):

    if np.sign(f(a)) == np.sign (f(b)):
        print ("Coordinate Invalid")
        return

    # find the middle value

    c = (a+b)/2

    # Recursive to find the root
    
    if np.abs(f(c)) < e:
        return c
    
    elif np.sign (f(a)) == np.sign(f(c)):
        return bisection (f, c, b, e)
    
    elif np.sign (f(b)) == np.sign (f(c)):
        return bisection (f, a, c, e)

### Root of Equation Code (Solution for Number 3)

In [18]:
# Polynomial Model from question number 1

poly_coeffs = np.polyfit(months, data, 5)
poly_model = np.poly1d(poly_coeffs)

# Function to find the root of equation (P(x) - 25000)

def f(x):
    return poly_model(x) - 25000

# First Derivative from Polynomial Model

poly_derivative = np.polyder(poly_model)

# First Derivative of Polynomial Model Function

def f_prime(x):
    return poly_derivative(x)

# Newton-Raphson Method

initial_guess = 130  # Initial guess (month that maybe produce 25000 bags)

# Using direct library

newton_root = math.ceil(newton(f, initial_guess, f_prime)) # 164.96
print(f'\nNewton-Raphson Method (Direct Method from Python Library): Month-{newton_root:.2f} (Therefore, the EGIER need to build a new warehouse at Month-{newton_root - 13:.2f})')

# Using custom Newton Raphson Function (based on the written above)

newton2_root = math.ceil(netwon_raphson(f, f_prime, initial_guess)) # 164.96
print(f'\nNewton-Raphson Method (Custom Function): Month-{newton2_root:.2f} (Therefore, the EGIER need to build a new warehouse at Month-{newton2_root - 13:.2f})')

# Bisection Method

a, b = -300, 300 # Interval

# Using direct library

bisection_root = math.ceil(bisect(f, a, b)) # 164.96
print(f'\nBisection Method (Direct Method from Python Library): Month-{bisection_root:.2f} (Therefore, the EGIER need to build a new warehouse at Month-{bisection_root - 13:.2f})')

# Using custom Bisection Function (based on the written above)

bisection2_root = math.ceil(bisection(f, a, b)) # 164.96
print(f'\nBisection Method (Custom Function): Month-{bisection2_root:.2f} (Therefore, the EGIER need to build a new warehouse at Month-{bisection2_root - 13:.2f})')

# Conclusion

print (f'\nAll of the method given the same answer value which is the Root of Equation is 165 so EGIER need to build a new warehouse at Month-152')

# Proofing that at Month-165 the bags production will be 25.000+ using Polynomial Model from Question Number 1

x_value = 165
y_value = poly_model(x_value)
print(f'\nY value when X = {x_value} is {y_value:.2f}')

x_value = 164
y_value = poly_model(x_value)
print(f'\nY value when X = {x_value} is {y_value:.2f}')


Newton-Raphson Method (Direct Method from Python Library): Month-165.00 (Therefore, the EGIER need to build a new warehouse at Month-152.00)

Newton-Raphson Method (Custom Function): Month-165.00 (Therefore, the EGIER need to build a new warehouse at Month-152.00)

Bisection Method (Direct Method from Python Library): Month-165.00 (Therefore, the EGIER need to build a new warehouse at Month-152.00)

Bisection Method (Custom Function): Month-165.00 (Therefore, the EGIER need to build a new warehouse at Month-152.00)

All of the method given the same answer value which is the Root of Equation is 165 so EGIER need to build a new warehouse at Month-152

Y value when X = 165 is 25016.99

Y value when X = 164 is 24579.60


### Conclusion

Based on the Polynomial Model that was created previously, it has been proven that in the 165th month, the number of bag production will be 25016.99. So that this number will exceed the existing storage capacity, it is necessary to build a new warehouse, 13 months earlier, namely the 152nd month. My assumption is to use Rounded up (ceiling), because in the 164th month, the amount of production has not exceeded the limit, so there is no need to start construction in the 151st month, but construction of the new warehouse can start in the 152nd month.

##### ==> EGIER need to start build a new warehouse in the 152nd month